### PR TITLE
Adding HaltMotion (Or new suggested name) to physics plugin

### PR DESCRIPTION
**Pull request by Lobotuerk**
Signed-off-by: Tomas Lorente <jtlorente@ekumenlabs.com>

# 🎉 New feature

## Summary
Adds the functionality to stop a model's joint's movement in other cases than the battery being 0. By adding a HaltMotion component to the model, its joints movement can be turned on/off. Also added some QoL to kineticEnergyMonitor plugin with profilers and some checking.

(Updates)
19acc8d ~~adds the capability to turn movement halting off if you only need the advice (and making it backward compatible).~~
4a9f551 Moved everything into GameLogicPlugin, just left here the halt component evaluation in physics and the QoL changes in kineticEnergyMonitor

## Test it
TO-DO tested it with SubT worlds for now

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: src/systems/kinetic_energy_monitor/KineticEnergyMonitor.hh [hunk 1]
```diff
@@ -97,9 +97,10 @@ namespace systems
       </plugin>
     </model>
   \endverbatim */
-  class IGNITION_GAZEBO_VISIBLE KineticEnergyMonitor:
-    public System,
+  class IGNITION_GAZEBO_VISIBLE KineticEnergyMonitor
+  : public System,
```
Comment (gonzodepedro) [inline src/systems/kinetic_energy_monitor/KineticEnergyMonitor.hh:101]: Unrelated change

Comment (Lobotuerk) [reply src/systems/kinetic_energy_monitor/KineticEnergyMonitor.hh:101]: I made some format changes, following the rest of the IGN plugins. Should that be another PR?

Comment (gonzodepedro) [reply src/systems/kinetic_energy_monitor/KineticEnergyMonitor.hh:101]: I'm ok with it, if it was intended.

File: include/ignition/gazebo/components/HaltMotion.hh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace components
+{
+  /// \brief A component used to indicate that an entity casts shadows
```
Comment (azeey) [inline include/ignition/gazebo/components/HaltMotion.hh:32]: nit: Can you update these docs?

File: src/systems/kinetic_energy_monitor/KineticEnergyMonitor.hh [hunk 3]
```diff
@@ -114,6 +115,10 @@ namespace systems
                            EntityComponentManager &_ecm,
                            EventManager &_eventMgr) final;
 
+    /// Documentation inherited
+    public: void Update(const UpdateInfo &_info,
```
Comment (azeey) [inline src/systems/kinetic_energy_monitor/KineticEnergyMonitor.hh:119]: Use of `Update` is discouraged. Only the physics system should use it. Can the functionality be added `PreUpdate`?

File: src/systems/kinetic_energy_monitor/KineticEnergyMonitor.cc [hunk 4]
```diff
@@ -161,12 +171,54 @@ void KineticEnergyMonitor::Configure(const Entity &_entity,
...
+
+  if (_info.paused)
+    return;
+
+  // Update Halt Motion component
+  auto *haltMotionComp =
+    _ecm.Component<components::HaltMotion>(
+      _ecm.ParentEntity(this->dataPtr->linkEntity));
+
+  if (this->dataPtr->haltMode &&
```
Comment (azeey) [inline src/systems/kinetic_energy_monitor/KineticEnergyMonitor.cc:206]: Should we check if `haltMotionComp` is nullptr before dereferencing it?

File: src/systems/kinetic_energy_monitor/KineticEnergyMonitor.cc [hunk 5]
```diff
@@ -124,6 +132,8 @@ void KineticEnergyMonitor::Configure(const Entity &_entity,
   ignmsg << "KineticEnergyMonitor publishing messages on "
     << "[" << topic << "]" << std::endl;
 
+  this->dataPtr->haltMode = sdfClone->Get<bool>("halt_mode", false).first;
```
Comment (azeey) [inline src/systems/kinetic_energy_monitor/KineticEnergyMonitor.cc:135]: Can you add documentation about `halt_mode` in `KineticEnergyMonitor.hh`?

**Review by azeey — COMMENTED** [review]
IMO, the kinetic energy monitor should not be concerned with halting a vehicle. If we adhere to the separation of concerns principle, I would think another System in SubT (eg. GameLogicPlugin) would monitor the output of the kinetic energy monitor and set the state of the `HaltMotion` component.

I also left some comments in case we decide to proceed with keeping that logic in the kinetic energy monitor.

**Comment by Lobotuerk** [comment]
> IMO, the kinetic energy monitor should not be concerned with halting a vehicle. If we adhere to the separation of concerns principle, I would think another System in SubT (eg. GameLogicPlugin) would monitor the output of the kinetic energy monitor and set the state of the `HaltMotion` component.

I do agree with moving the HaltMotion logic into the GameLogicPlugin, but that would require adding A PreUpdate there. @nkoenig Do you have a strong feeling for either of them?

**Comment by nkoenig** [comment]
> IMO, the kinetic energy monitor should not be concerned with halting a vehicle. If we adhere to the separation of concerns principle, I would think another System in SubT (eg. GameLogicPlugin) would monitor the output of the kinetic energy monitor and set the state of the `HaltMotion` component.
> 
> I also left some comments in case we decide to proceed with keeping that logic in the kinetic energy monitor.

Agreed. @Lobotuerk , can you remove the logic for halting motion from the Kinetic Energy Monitor and place it in the SubT Game Logic Plugin?

**Comment by nkoenig** [comment]
> 
> I do agree with moving the HaltMotion logic into the GameLogicPlugin, but that would require adding A PreUpdate there. @nkoenig Do you have a strong feeling for either of them?

Adding a PreUpdate to the GameLogicPlugin sounds good.

**Comment by Lobotuerk** [comment]
@azeey When moving the halting logic to GameLogic, should I leave the HaltMotion component creation inside KineticEnergyMonitor?

File: src/systems/kinetic_energy_monitor/KineticEnergyMonitor.hh (hunk 3 above)
Comment (Lobotuerk) [reply src/systems/kinetic_energy_monitor/KineticEnergyMonitor.hh]: Moved to https://github.com/osrf/subt/pull/858

File: src/systems/kinetic_energy_monitor/KineticEnergyMonitor.cc (hunk 4 above)
Comment (Lobotuerk) [reply src/systems/kinetic_energy_monitor/KineticEnergyMonitor.cc]: Moved to https://github.com/osrf/subt/pull/858 and fixed

File: src/systems/kinetic_energy_monitor/KineticEnergyMonitor.cc (hunk 5 above)
Comment (Lobotuerk) [reply src/systems/kinetic_energy_monitor/KineticEnergyMonitor.cc:132]: Added on  f5c472e

File: include/ignition/gazebo/components/HaltMotion.hh (hunk 2 above)
Comment (Lobotuerk) [reply include/ignition/gazebo/components/HaltMotion.hh]: Updated in 67c262b

**Comment by azeey** [comment]
> @azeey When moving the halting logic to GameLogic, should I leave the HaltMotion component creation inside KineticEnergyMonitor?

IMO, We don't need to modify `KineticEnergyMonitor` at all. All the logic can be in `GameLogic` and `Physics`.

**Review by azeey — APPROVED** [review]
LGTM! It would be great if you can add a test showing that a vehicle stops once the `HaltMotion` component is set.

**Review by chapulina — COMMENTED** [review]
I'm wondering if this functionality couldn't be achieved using `components::JointVelocityReset` / `components::JointPositionReset`?

**Review by nkoenig — APPROVED** [review]

**Comment by Lobotuerk** [comment]
> I'm wondering if this functionality couldn't be achieved using `components::JointVelocityReset` / `components::JointPositionReset`?

It's better for a long time halt to simply stop the physics layer before applying forces, than trying to reset the joint every single timestep
